crates/optimism/node/src/node.rs

```diff
@@ -286,12 +291,15 @@ where
         ctx: &BuilderContext<Node>,
         pool: Pool,
     ) -> eyre::Result<NetworkHandle> {
-        let Self { disable_txpool_gossip } = self;
+        let Self { disable_txpool_gossip, disable_discovery_v4 } = self;
 
         let args = &ctx.config().network;
 
-        let network_builder = ctx
-            .network_config_builder()?
+        let mut network_config_builder = ctx.network_config_builder()?;
+        if disable_discovery_v4 || args.discovery.disable_discovery {
+            network_config_builder = network_config_builder.disable_discv4_discovery();
+        }
+        let network_builder = network_config_builder
```

Comment:
style nit, you can use the `apply` function to keep the builder style chain and perform this check in the closure